util/chunk/row_container.go

```diff
@@ -28,149 +29,175 @@ import (
+}
 
 // NumRow returns the number of rows in the container
 func (c *RowContainer) NumRow() int {
-	if c.AlreadySpilled() {
-		return c.recordsInDisk.Len()
+	c.m.RLock()
+	defer c.m.RUnlock()
+	if c.alreadySpilled() {
+		return c.m.recordsInDisk.Len()
 	}
-	return c.records.Len()
+	return c.m.records.Len()
 }
 
 // NumRowsOfChunk returns the number of rows of a chunk in the ListInDisk.
 func (c *RowContainer) NumRowsOfChunk(chkID int) int {
-	if c.AlreadySpilled() {
-		return c.recordsInDisk.NumRowsOfChunk(chkID)
+	c.m.RLock()
+	defer c.m.RUnlock()
+	if c.alreadySpilled() {
+		return c.m.recordsInDisk.NumRowsOfChunk(chkID)
 	}
-	return c.records.NumRowsOfChunk(chkID)
+	return c.m.records.NumRowsOfChunk(chkID)
 }
 
 // NumChunks returns the number of chunks in the container.
 func (c *RowContainer) NumChunks() int {
-	if c.AlreadySpilled() {
-		return c.recordsInDisk.NumChunks()
+	c.m.RLock()
+	defer c.m.RUnlock()
+	if c.alreadySpilled() {
+		return c.m.recordsInDisk.NumChunks()
 	}
-	return c.records.NumChunks()
+	return c.m.records.NumChunks()
 }
 
 // Add appends a chunk into the RowContainer.
 func (c *RowContainer) Add(chk *Chunk) (err error) {
-	if c.AlreadySpilled() {
-		err = c.recordsInDisk.Add(chk)
-	} else {
-		c.records.Add(chk)
-		if atomic.LoadUint32(&c.exceeded) != 0 {
-			if c.onExceededCallback != nil {
-				c.onExceededCallback(c)
-			}
-			err = c.spillToDisk()
-			if err != nil {
-				return err
-			}
-			atomic.StoreUint32(&c.spilled, 1)
+	c.m.RLock()
+	defer c.m.RUnlock()
+	if c.alreadySpilled() {
+		if c.m.spillError != nil {
+			return c.m.spillError
 		}
+		err = c.m.recordsInDisk.Add(chk)
+	} else {
+		c.m.records.Add(chk)
 	}
 	return
 }
 
 // AppendRow appends a row to the RowContainer, the row is copied to the RowContainer.
 func (c *RowContainer) AppendRow(row Row) (RowPtr, error) {
-	if c.AlreadySpilled() {
+	if c.alreadySpilled() {
 		return RowPtr{}, errors.New("ListInDisk don't support AppendRow")
 	}
-	return c.records.AppendRow(row), nil
+	return c.m.records.AppendRow(row), nil
 }
 
 // AllocChunk allocates a new chunk from RowContainer.
 func (c *RowContainer) AllocChunk() (chk *Chunk) {
-	return c.records.allocChunk()
+	return c.m.records.allocChunk()
 }
 
 // GetChunk returns chkIdx th chunk of in memory records.
-func (c *RowContainer) GetChunk(chkIdx int) *Chunk {
-	return c.records.GetChunk(chkIdx)
-}
-
-// GetList returns the list of in memory records.
-func (c *RowContainer) GetList() *List {
-	return c.records
+func (c *RowContainer) GetChunk(chkIdx int) (*Chunk, error) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	if !c.alreadySpilled() {
+		return c.m.records.GetChunk(chkIdx), nil
+	}
+	chk := NewChunkWithCapacity(c.m.records.FieldTypes(), c.m.records.maxChunkSize)
+	offsets := c.m.recordsInDisk.offsets[chkIdx]
+	for rowIdx := range offsets {
+		row, err := c.m.recordsInDisk.GetRow(RowPtr{ChkIdx: uint32(chkIdx), RowIdx: uint32(rowIdx)})
+		if err != nil {
+			return chk, err
+		}
+		chk.AppendRow(row)
+	}
```

Comment:
Would it be better to put this logic into a method in `recordsInDisk.GetChunk` so it's easy for us to maintain if we refactor or optimize the internal data format of `recordsInDisk` in future?

Reply:
ok